src/inmanta/execute/dataflow/datatrace.py

```diff
@@ -0,0 +1,241 @@
+"""
+    Copyright 2020 Inmanta
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+    Contact: code@inmanta.com
+"""
+
+from itertools import chain
+from typing import TYPE_CHECKING, Iterable, List, Optional
+
+from inmanta.ast import Locatable, NotFoundException
+from inmanta.ast.statements import Statement
+from inmanta.execute.dataflow import (
+    AssignableNode,
+    AssignableNodeReference,
+    Assignment,
+    AttributeNode,
+    AttributeNodeReference,
+    DataflowGraph,
+    Equivalence,
+    InstanceNode,
+    InstanceNodeReference,
+    VariableNodeReference,
+)
+from inmanta.execute.runtime import Instance
+
+if TYPE_CHECKING:
+    pass
+
+
+class DataTraceRenderer:
+    @classmethod
+    def _prefix_line(cls, prefix: str, line: str) -> str:
+        """
+            Prefixes a line.
+        """
+        if line == "":
+            return prefix.rstrip()
+        return prefix + line
+
+    @classmethod
+    def _prefix(cls, prefix: str, lines: Iterable[str]) -> List[str]:
+        """
+            Prefixes lines.
+        """
+        return [cls._prefix_line(prefix, line) for line in lines]
+
+    @classmethod
+    def _branch(cls, lines: List[str], last: Optional[bool] = False) -> List[str]:
+        """
+            Renders a branch in the tree.
+
+            :param last: True iff this is the last branch on this level.
+        """
+        if len(lines) == 0:
+            return []
+        branch_prefix: str = ("└" if last else "├") + "── "
+        block_prefix: str = (" " if last else "│") + " " * 3
+        result: List[str] = cls._prefix(branch_prefix, lines[0:1])
+        result += cls._prefix(block_prefix, lines[1:])
+        return result
+
+    @classmethod
+    def _indent(cls, lines: Iterable[str]) -> List[str]:
+        """
+            Indents lines.
+        """
+        return cls._prefix(" " * 4, lines)
+
+    @classmethod
+    def _render_implementation_context(cls, context: DataflowGraph) -> List[str]:
+        """
+            Renders information about the dynamic implementation context, if it exists.
+        """
+        try:
+            context.resolver.lookup("self")
+        except NotFoundException:
+            return []
+        result: List[str] = []
+        var_node: AssignableNodeReference = context.resolver.get_dataflow_node("self")
+        if (
```

Comment:
This if statement could be an assertion I assume? (in that it should always be true?)

Reply:
Alas not, take below model as an example:
```
# bad practice, but model developer doesn't care
self = "myself"
x = 42
```
In this context there is a `self` variable but we're not in an implementation. This if statements guards against this false positive. It doesn't prevent all false positives though.
I'm not really satisfied with this method for implementation context detection, but I'm not sure how to improve it. When adding support for for loops this definitely needs to be handled differently though.

Reply:
Perhaps a `DataflowGraph` instance could keep a reference to the closest enveloping implementation context? When an `Instance` creates a new `DataflowGraph` instance it could then register itself.

Reply:
This, paired with below comment of which implementation, is a limitation of the current graph implementation. This should definitely be resolved but I'm not sure if it belongs in this PR or if I should create a separate issue. What do you think?
(the current trace doesn't tell you which implementation but it does show a line number)

Reply:
The resolution doesn't belong here. 
The comment why this is here should go into the code.

Reply:
So:
- in this PR add a comment
- create a ticket for resolving the discussed limitation

Correct?

Reply:
yes